front/pages/api/w/[wId]/vaults/[vId]/data_sources/index.ts

```diff
@@ -0,0 +1,654 @@
+    return apiError(req, res, {
+      status_code: 400,
+      api_error: {
+        type: "invalid_request_error",
+        message: "The data source name cannot start with `managed-`.",
+      },
+    });
+  }
+  if (!isDataSourceNameValid(name)) {
+    return apiError(req, res, {
+      status_code: 400,
+      api_error: {
+        type: "invalid_request_error",
+        message:
+          "Data source names must only contain letters, numbers, and the characters `._-`, and cannot be empty.",
+      },
+    });
+  }
+  const dataSources = await DataSourceResource.listByWorkspace(auth);
+  if (
+    plan.limits.dataSources.count != -1 &&
+    dataSources.length >= plan.limits.dataSources.count
+  ) {
+    return apiError(req, res, {
+      status_code: 401,
+      api_error: {
+        type: "plan_limit_error",
+        message: "Your plan does not allow you to create managed data sources.",
+      },
+    });
+  }
+
+  const dataSourceEmbedder =
+    owner.defaultEmbeddingProvider ?? DEFAULT_EMBEDDING_PROVIDER_ID;
+  const embedderConfig = EMBEDDING_CONFIGS[dataSourceEmbedder];
+  const coreAPI = new CoreAPI(config.getCoreAPIConfig(), logger);
+
+  const dustProject = await coreAPI.createProject();
+  if (dustProject.isErr()) {
+    return apiError(req, res, {
+      status_code: 500,
+      api_error: {
+        type: "internal_server_error",
+        message: `Failed to create internal project for the data source.`,
+        data_source_error: dustProject.error,
+      },
+    });
+  }
+
+  const dustDataSource = await coreAPI.createDataSource({
+    projectId: dustProject.value.project.project_id.toString(),
+    dataSourceId: name,
+    config: {
+      qdrant_config: {
+        cluster: DEFAULT_QDRANT_CLUSTER,
+        shadow_write_cluster: null,
+      },
+      embedder_config: {
+        embedder: {
+          max_chunk_size: embedderConfig.max_chunk_size,
+          model_id: embedderConfig.model_id,
+          provider_id: embedderConfig.provider_id,
+          splitter_id: embedderConfig.splitter_id,
+        },
+      },
+    },
+    credentials: dustManagedCredentials(),
+  });
+
+  if (dustDataSource.isErr()) {
+    return apiError(req, res, {
+      status_code: 500,
+      api_error: {
+        type: "internal_server_error",
+        message: "Failed to create the data source.",
+        data_source_error: dustDataSource.error,
+      },
+    });
+  }
+
+  const dataSource = await DataSourceResource.makeNew(
+    {
+      name,
+      description,
+      dustAPIProjectId: dustProject.value.project.project_id.toString(),
+      workspaceId: owner.id,
+      assistantDefaultSelected: false,
+      editedByUserId: user.id,
+    },
+    vault
+  );
+
+  const dataSourceView =
+    await DataSourceViewResource.createViewInVaultFromDataSourceIncludingAllDocuments(
+      dataSource.vault,
+      dataSource
+    );
+
+  const dataSourceType = dataSource.toJSON();
+
+  res.status(201).json({
+    dataSource: dataSourceType,
```

Comment:
can we just dou dataSource.toJSON()?

Reply:
Yes, I used a var because we use it twice so I thought one serialization was better than two. Adding a commit then

Reply:
The main reason is that the var is quite ill named imho. V8 probably does that for you also. But happy to keep a var with a better name but really don't think double serialization an issue if it makes the code more readable 👍